### PR TITLE
feat: ServerCookieEncoder and ServerCookieDecoder

### PR DESCRIPTION
**Pull request by sdelamo**
reverts https://github.com/micronaut-projects/micronaut-core/pull/10364

This pull request adds an API `ClientCookieEncoder`, `SeverCookieEncoder` and `ServerCookieDecoder` so that other HTTP implementations, for example, Micronaut AWS Lambda, can altogether remove the dependency on Netty.

### CODE REVIEW

File: http-netty/src/main/java/io/micronaut/http/netty/cookies/NettyLaxClientCookieEncoder.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import io.micronaut.http.cookie.Cookie;
+
+import static io.netty.handler.codec.http.cookie.ClientCookieEncoder.LAX;
+
+/**
+ * {@link ClientCookieEncoder} implementation backed on Netty's {@link io.netty.handler.codec.http.cookie.ClientCookieEncoder#LAX}.
+ * @author Sergio del Amo
+ * @since 4.3.0
+ */
+public class NettyLaxClientCookieEncoder implements ClientCookieEncoder {
```
Comment (graemerocher) [inline http-netty/src/main/java/io/micronaut/http/netty/cookies/NettyLaxClientCookieEncoder.java]: can this be `final`? Should it be marked `@Internal`?

File: http-netty/src/main/java/io/micronaut/http/netty/cookies/NettyLaxServerCookieDecoder.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import java.util.List;
+import static io.netty.handler.codec.http.cookie.ServerCookieDecoder.LAX;
+
+/**
+ * @author Sergio del Amo
+ * @since 4.3.0
+ */
+@Internal
+public class NettyLaxServerCookieDecoder implements ServerCookieDecoder {
```
Comment (graemerocher) [inline http-netty/src/main/java/io/micronaut/http/netty/cookies/NettyLaxServerCookieDecoder.java]: can this be `final`?

File: http-netty/src/main/java/io/micronaut/http/netty/cookies/NettyServerCookieEncoder.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.List;
+import static io.netty.handler.codec.http.cookie.ServerCookieEncoder.STRICT;
+
+/**
+ * {@link ServerCookieEncoder} implementation backed by {@link io.netty.handler.codec.http.cookie.ServerCookieEncoder}.
+ * @author Sergio del Amo
+ * @since 4.3.0
+ */
+@Internal
+public class NettyServerCookieEncoder implements ServerCookieEncoder {
```
Comment (graemerocher) [inline http-netty/src/main/java/io/micronaut/http/netty/cookies/NettyServerCookieEncoder.java]: can this be `final`?

File: http/src/main/java/io/micronaut/http/cookie/ClientCookieEncoder.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import io.micronaut.core.io.service.ServiceDefinition;
+import io.micronaut.core.io.service.SoftServiceLoader;
+
+/**
+ *
+ * @author Sergio del Amo
+ * @since 4.3.0
+ */
+@FunctionalInterface
+public interface ClientCookieEncoder {
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/cookie/ClientCookieEncoder.java:30]: add more javadoc to type describing what this is

File: http/src/main/java/io/micronaut/http/cookie/ClientCookieEncoder.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+     * The default {@link ServerCookieEncoder} instance.
+     */
+    ClientCookieEncoder INSTANCE = SoftServiceLoader
+            .load(ClientCookieEncoder.class)
+            .firstOr("io.micronaut.http.cookie.DefaultClientCookieEncoder", ClientCookieEncoder.class.getClassLoader())
+            .map(ServiceDefinition::load)
+            .orElse(null);
+
+    @NonNull
+    String encode(@NonNull Cookie cookie);
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/cookie/ClientCookieEncoder.java:48]: missing javadoc

File: http/src/main/java/io/micronaut/http/cookie/CookieComparator.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.util.Comparator;
+
+/**
+ * {@link Comparator} for {@link Cookie}.
+ * Implementation mimics implementation of Netty {@link io.netty.handler.codec.http.cookie.DefaultCookie#compareTo(io.netty.handler.codec.http.cookie.Cookie)}.
+ * @author Sergio del Amo
+ * @since 4.3.0
+ */
+public class CookieComparator implements Comparator<Cookie> {
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/cookie/CookieComparator.java]: can this be `final`?

File: http/src/main/java/io/micronaut/http/cookie/CookieHttpCookieAdapter.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+import java.net.HttpCookie;
+import java.util.Optional;
+
+/**
+ * Adapt from {@link HttpCookie} to {@link Cookie}.
+ * @author Sergio del Amo
+ * @since 4.3.0
+ */
+@Internal
+public class CookieHttpCookieAdapter implements Cookie {
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/cookie/CookieHttpCookieAdapter.java]: can this be `final`? Does it need to be `public`?

File: http/src/main/java/io/micronaut/http/cookie/DefaultClientCookieEncoder.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+package io.micronaut.http.cookie;
+
+import io.micronaut.core.annotation.Internal;
+
+/**
+ * @author Sergio del Amo
+ * @since 4.3.0
+ */
+@Internal
+public class DefaultClientCookieEncoder implements ClientCookieEncoder {
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/cookie/DefaultClientCookieEncoder.java]: can this be `final`. Does it need to be `public`?

File: http/src/main/java/io/micronaut/http/cookie/DefaultServerCookieDecoder.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+import java.net.HttpCookie;
+import java.util.List;
+
+/**
+ * {@link ServerCookieDecoder} implementation backed by {@link HttpCookie#parse(String)}.
+ *
+ * @author Sergio del Amo
+ * @since 4.3.0
+ */
+public class DefaultServerCookieDecoder implements ServerCookieDecoder {
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/cookie/DefaultServerCookieDecoder.java]: Can this be `final` and `@Internal`? Does it need to be public?

File: http/src/main/java/io/micronaut/http/cookie/DefaultServerCookieEncoder.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Implementation of {@link ServerCookieEncoder} which uses {@link HttpCookie} to encode cookies.
+ * @author Sergio del Amo
+ * @since 4.4.0
+ */
+@Internal
+public class DefaultServerCookieEncoder implements ServerCookieEncoder {
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/cookie/DefaultServerCookieEncoder.java]: can this be `final`? Does it need to be `public`?

File: http/src/main/java/io/micronaut/http/cookie/HttpCookieFactory.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+package io.micronaut.http.cookie;
+
+import java.net.HttpCookie;
+
+/**
+ * {@link CookieFactory} implementation backed by {@link java.net.HttpCookie}.
+ * @author Sergio del Amo
+ * @since 4.3.0
+ */
+public class HttpCookieFactory implements CookieFactory {
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/cookie/HttpCookieFactory.java]: can this be `final`? Does it need to be `public`? Should it be `@Internal`?

File: http/src/main/java/io/micronaut/http/cookie/ServerCookieDecoder.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+@FunctionalInterface
+public interface ServerCookieDecoder {
+    /**
+     * The default {@link ServerCookieDecoder} instance.
+     */
+    ServerCookieDecoder INSTANCE = SoftServiceLoader
+            .load(ServerCookieDecoder.class)
+            .firstOr("io.micronaut.http.cookie.DefaultServerCookieDecoder", ServerCookieDecoder.class.getClassLoader())
+            .map(ServiceDefinition::load)
+            .orElse(null);
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/cookie/ServerCookieDecoder.java:38]: we should probably initialise these at build time for Graal

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: http/src/main/java/io/micronaut/http/cookie/Cookie.java [hunk 13]
```diff
@@ -24,13 +24,49 @@
...
  *
  * @author Graeme Rocher
  * @since 1.0
  */
 public interface Cookie extends Comparable<Cookie>, Serializable {
 
+    /**
+     * @see <a href="https://tools.ietf.org/html/rfc6265#section-4.1.1">The Secure Attribute</a>.
+     */
+    String ATTRIBUTE_SECURE = "Secure";
```
Comment (yawkat) [inline http/src/main/java/io/micronaut/http/cookie/Cookie.java:38]: do we need these?

File: http/src/main/java/io/micronaut/http/cookie/DefaultServerCookieEncoder.java [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+     * {@link HttpCookie} default value for max age is -1 and io.netty.handler.codec.http.cookie.DefaultCookie is Long.MIN_VALUE.
+     * @param cookie Cookie
+     * @return true if the cookie's max age is set.
+     */
+    private boolean isMaxAgeSet(Cookie cookie) {
+        return cookie.getMaxAge() != -1 && cookie.getMaxAge() != Long.MIN_VALUE;
+    }
+
+    private static String expires(Long maxAgeSeconds) {
+        LocalDateTime localDateTime = LocalDateTime.now(GMT_ZONE).plusSeconds(maxAgeSeconds);
```
Comment (yawkat) [inline http/src/main/java/io/micronaut/http/cookie/DefaultServerCookieEncoder.java]: just use ZoneOffset.UTC

File: http/src/test/java/io/micronaut/http/cookie/CookeFactoryTest.java [hunk 15]
```diff
@@ -0,0 +1,13 @@
+package io.micronaut.http.cookie;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
+
+class CookeFactoryTest {
```
Comment (yawkat) [inline http/src/test/java/io/micronaut/http/cookie/CookeFactoryTest.java]: typo

**Review by yawkat — APPROVED** [review]

File: http/src/main/java/io/micronaut/http/cookie/Cookie.java (hunk 13 above)
Comment (sdelamo) [reply http/src/main/java/io/micronaut/http/cookie/Cookie.java:38]: Yes, I am using them in the encoder.

File: http/src/test/java/io/micronaut/http/cookie/CookeFactoryTest.java (hunk 15 above)
Comment (sdelamo) [reply http/src/test/java/io/micronaut/http/cookie/CookeFactoryTest.java]: good catch. fixed via https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/fa3ffe5dc62408d8bd064499291939fa4695a4bb

File: http/src/main/java/io/micronaut/http/cookie/DefaultServerCookieEncoder.java (hunk 10 above)
Comment (sdelamo) [reply http/src/main/java/io/micronaut/http/cookie/DefaultServerCookieEncoder.java]: > can this be final? 

https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/a8d3204d9c512fcc50950358fd14bf25905b6bbe

> Does it need to be public?

Yes, it needs to be public.

File: http/src/main/java/io/micronaut/http/cookie/DefaultClientCookieEncoder.java (hunk 8 above)
Comment (sdelamo) [reply http/src/main/java/io/micronaut/http/cookie/DefaultClientCookieEncoder.java]: > can this be final? 

https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/a8d3204d9c512fcc50950358fd14bf25905b6bbe

> Does it need to be public?

Yes, it needs to be public.

File: http/src/main/java/io/micronaut/http/cookie/CookieHttpCookieAdapter.java (hunk 7 above)
Comment (sdelamo) [reply http/src/main/java/io/micronaut/http/cookie/CookieHttpCookieAdapter.java]: > can this be final? 

https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/a8d3204d9c512fcc50950358fd14bf25905b6bbe

> Does it need to be public?

[I removed the public modifier](https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/f4a6f90d37608c74172a85f2d9583f20f09cc925)

File: http/src/main/java/io/micronaut/http/cookie/CookieComparator.java (hunk 6 above)
Comment (sdelamo) [reply http/src/main/java/io/micronaut/http/cookie/CookieComparator.java]: > can this be final? 

https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/a8d3204d9c512fcc50950358fd14bf25905b6bbe

File: http-netty/src/main/java/io/micronaut/http/netty/cookies/NettyServerCookieEncoder.java (hunk 3 above)
Comment (sdelamo) [reply http-netty/src/main/java/io/micronaut/http/netty/cookies/NettyServerCookieEncoder.java]: > can this be final? 

https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/a8d3204d9c512fcc50950358fd14bf25905b6bbe

File: http-netty/src/main/java/io/micronaut/http/netty/cookies/NettyLaxServerCookieDecoder.java (hunk 2 above)
Comment (sdelamo) [reply http-netty/src/main/java/io/micronaut/http/netty/cookies/NettyLaxServerCookieDecoder.java]: > can this be final? 

https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/a8d3204d9c512fcc50950358fd14bf25905b6bbe

File: http-netty/src/main/java/io/micronaut/http/netty/cookies/NettyLaxClientCookieEncoder.java (hunk 1 above)
Comment (sdelamo) [reply http-netty/src/main/java/io/micronaut/http/netty/cookies/NettyLaxClientCookieEncoder.java]: > can this be `final`? 

https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/a8d3204d9c512fcc50950358fd14bf25905b6bbe

>  Should it be marked `@Internal`?

https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/ce14d1e942177abd7c82c437888b4d283759009a

File: http/src/main/java/io/micronaut/http/cookie/HttpCookieFactory.java (hunk 11 above)
Comment (sdelamo) [reply http/src/main/java/io/micronaut/http/cookie/HttpCookieFactory.java]: > can this be `final`? 

https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/a8d3204d9c512fcc50950358fd14bf25905b6bbe

> can this be `@Internal`? 
https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/ce14d1e942177abd7c82c437888b4d283759009a

> Does it need to be public?

Yes, it needs to be public.

File: http/src/main/java/io/micronaut/http/cookie/DefaultServerCookieDecoder.java (hunk 9 above)
Comment (sdelamo) [reply http/src/main/java/io/micronaut/http/cookie/DefaultServerCookieDecoder.java]: > can this be `final`? 

https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/a8d3204d9c512fcc50950358fd14bf25905b6bbe

> can this be `@Internal`? 
https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/ce14d1e942177abd7c82c437888b4d283759009a

> Does it need to be public?

Yes, it needs to be public.

File: http/src/main/java/io/micronaut/http/cookie/ClientCookieEncoder.java (hunk 4 above)
Comment (sdelamo) [reply http/src/main/java/io/micronaut/http/cookie/ClientCookieEncoder.java:30]: https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/648f73c40363b3219a6db8ffb6af6a9e61944a98

File: http/src/main/java/io/micronaut/http/cookie/ClientCookieEncoder.java (hunk 5 above)
Comment (sdelamo) [reply http/src/main/java/io/micronaut/http/cookie/ClientCookieEncoder.java:48]: https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/648f73c40363b3219a6db8ffb6af6a9e61944a98

File: http/src/main/java/io/micronaut/http/cookie/ServerCookieDecoder.java (hunk 12 above)
Comment (sdelamo) [reply http/src/main/java/io/micronaut/http/cookie/ServerCookieDecoder.java:38]: the `nativeTest` seems to pass. 

Do you want me to create a `native-image.properties` in the `http` module with content: 

```
Args = --initialize-at-build-time=io.micronaut.http.cookie
```

File: http/src/main/java/io/micronaut/http/cookie/DefaultServerCookieEncoder.java (hunk 14 above)
Comment (sdelamo) [reply http/src/main/java/io/micronaut/http/cookie/DefaultServerCookieEncoder.java]: done via https://github.com/micronaut-projects/micronaut-core/pull/10402/commits/42e620d29b51516c8b0051cc0432fa4d9efeb627

File: http/src/main/java/io/micronaut/http/cookie/ServerCookieDecoder.java (hunk 12 above)
Comment (graemerocher) [reply http/src/main/java/io/micronaut/http/cookie/ServerCookieDecoder.java:38]: if it doesn't break anything then makes sense

**Review by graemerocher — APPROVED** [review]

File: http/src/main/java/io/micronaut/http/cookie/ServerCookieDecoder.java (hunk 12 above)
Comment (sdelamo) [reply http/src/main/java/io/micronaut/http/cookie/ServerCookieDecoder.java:38]: Do you mean I should add the `native-image.properties` file and initialize the package `io.micronaut.http.cookie` at build time or leave it as is and don't add any native image configuration?

**Pull request merged**